### PR TITLE
Ignore alias option

### PR DESCRIPTION
**Pull request by bendlas**
Locking all aliases may not be desirable for every project,
e.g. for development-only aliases with local paths, and such.

This PR adds the ability to ignore aliases, similar to ignoring deps files.

For some use cases, the existing ignore-deps capability might be used to achive the same result as with ignored aliases, but with some tools, that's inconvenient. E.g. shadow-cljs doesn't seem to implement support for multiple overlaid deps files.

I'm still working out how to test this and add some error handling to the options parser, but wanted to get some feedback in the meantime about the options interface and if you want this PR at all.

### CODE REVIEW

File: src/cljnix/core.clj [hunk 1]
```diff
@@ -391,6 +394,21 @@
...
+(defn parse-options [args]
+  (loop [[arg & next-args] args
+         options {:lein? false
+                  :aliases-ignore []
+                  :deps-ignore []}]
+    (case arg
+      nil              options
+      "--lein"         (recur next-args
+                              (assoc options :lein? true))
+      "--ignore-alias" (recur (next next-args)
```
Comment (bendlas) [inline src/cljnix/core.clj:406]: How do you feel about the naming and having multiple arguments of this vs single argument with comma-separated values?

File: src/cljnix/core.clj [hunk 2]
```diff
@@ -391,6 +394,21 @@
...
+                              (update options :aliases-ignore conj
+                                      (keyword (first next-args))))
+      (recur next-args
+             (update options :deps-ignore conj arg)))))
```
Comment (bendlas) [inline src/cljnix/core.clj:410]: Seems like we're painted into a bit of a corner by having varargs for `:deps-ignore` in the public interface. Maybe we could add a task `nix run .#deps-lock2 -- --deps-ignore a/b,c/d --alias-ignore al1,al2`

**Comment by jlesquembre** [comment]
Thanks for this PR, that's the direction I want to go.

My plan was to re-implement the CLI to be more granular about dependencies.  My idea is to have the following flags:

- To include/exclude `*.edn` files: `--deps-include` and `--deps-exclude`. Those should support glob patterns.
- To include/exclude aliases: `--alias-include` and `--alias-exclude`  (in both cases, `exclude` options takes preference over `include`)
- `--bb`: Include `bb.edn` files, Probably in the end it will be an alias for `--deps-include **/bb.edn`
- `--lein`: Include lein files (already implemented)

Without include/exclude options, we include everything.

I'd like to hear your toughs about the flags. Anything to add or change?

I also thought about adding a library like [tools.cli](https://github.com/clojure/tools.cli) or [babashka.cli](https://github.com/babashka/cli), but maybe the way you are implementing it with `loop-case` is enough, what do you think?

I'm answering your other questions separately.

My question is how you want to proceed. As an initial implementation, it's good, but I want to add more functionality. For me, it's ok if you want to work on it. Depending on how much time you want to invest, you could implement it and get it merged to main, or we could merge a subset of the functionality into a branch, and I can take it from there.

File: src/cljnix/core.clj (hunk 1 above)
Comment (jlesquembre) [reply src/cljnix/core.clj:406]: I don't have a preference, pick the one you prefer.

File: src/cljnix/core.clj (hunk 2 above)
Comment (jlesquembre) [reply src/cljnix/core.clj:410]: Agree, I don't like having varargs, it was a quick implementation I wanted to refactor later, but you know, I never did it. In this case I think is ok to introduce a breaking change, the current interface is not good.

**Comment by bendlas** [comment]
> My question is how you want to proceed. As an initial implementation, it's good, but I want to add more functionality. For me, it's ok if you want to work on it. Depending on how much time you want to invest, you could implement it and get it merged to main, or we could merge a subset of the functionality into a branch, and I can take it from there.

Thanks to how flakes do, I'm already putting this to work for my project, and so I've got no problem if you feel like taking over. I'm also happy to help with implementing after we nail down the details.

> * To include/exclude `*.edn` files: `--deps-include` and `--deps-exclude`. Those should support glob patterns.
> * To include/exclude aliases: `--alias-include` and `--alias-exclude`  (in both cases, `exclude` options takes preference over `include`)
> * `--bb`: Include `bb.edn` files, Probably in the end it will be an alias for `--deps-include **/bb.edn`
> * `--lein`: Include lein files (already implemented)
> 
> Without include/exclude options, we include everything.
> 
> I'd like to hear your toughs about the flags. Anything to add or change?

I'm wondering about the ergonomics of wildcards: Usually, wildcards would be expanded by the shell, so in order to pass them to `deps-lock`, they'd need to be quoted like `--deps-include '**/bb.edn'` and if user forgot to quote the wildcard, results would probably be confusing.

OTOH, if the interface was designed around shell wildcard expansion, it would also result in non-standard interface: `--deps-include bb1.edn bb2.edn`

Though does clj-nix really need wildcard support? What use cases are there, given that the deps-lock invokation could easily be generated by nix into a command for the flake using clj-nix?

> I also thought about adding a library like [tools.cli](https://github.com/clojure/tools.cli) or [babashka.cli](https://github.com/babashka/cli), but maybe the way you are implementing it with `loop-case` is enough, what do you think?

Using a library seems like a good idea to me. Standard interface with generated help and all ...

**Comment by jlesquembre** [comment]
Ok, you convinced me about wildcard support, let's leave it out. Anyways projects don't have that many deps files, so probably it's going to cause trouble for almost no benefit. 

> Using a library seems like a good idea to me. Standard interface with generated help and all ...

Any suggestion? I'm not familiar with any clojure CLI library, I like [tools.cli](https://github.com/clojure/tools.cli) or [babashka.cli](https://github.com/babashka/cli) because don't have any dependencies, but I'm open to suggestions.

**Comment by bendlas** [comment]
> Any suggestion? I'm not familiar with any clojure CLI library, I like [tools.cli](https://github.com/clojure/tools.cli) or [babashka.cli](https://github.com/babashka/cli) because don't have any dependencies, but I'm open to suggestions.

I haven't formed an opinion on CLI libs so far either. Maybe go with tools.cli by default? Try with both and compare outcomes?

**Comment by jlesquembre** [comment]
I'm thinking about merging this PR, and continuing working on the CLI on another PR, any objections?

Regarding CLI options, we can have:
- deps-include
- deps-exclude
- alias-include
- alias-exclude
- bb (Include bb.edn deps, false by default)
- lein (Include lein deps, false by default)

No wildcards, `deps-include` / `deps-exclude` and `alias-include` / `alias-exclude` are mutually exclusive. Without any options, all deps.edn / aliases are included. Makes sense to you? Any suggestions about the option names?

Something I forgot to mention, in the current version we also have some other flags, like `--jar` or `--patch-git-sha`. I don't like to expose those flags, I'll move those somewhere else.

About the CLI libraries, I played a bit with both, no strong opinions, but I'm liking `babashka.cli` a bit more, probably I'll go with that one.

**Comment by bendlas** [comment]
> I'm thinking about merging this PR, and continuing working on the CLI on another PR, any objections?

No objections.

> I'm liking `babashka.cli` a bit more, probably I'll go with that one.

LGTM

> No wildcards, deps-include / deps-exclude and alias-include / alias-exclude are mutually exclusive. Without any options, all deps.edn / aliases are included. Makes sense to you?

"all deps.edn" as in the current behavior? All files named `deps.edn` in subdirectories?

> Any suggestions about the option names?

Your proposals seem fine to me.

**Comment by jlesquembre** [comment]
> "all deps.edn" as in the current behavior? All files named deps.edn in subdirectories?

Yes, as in the current behavior, all `deps.edn` in all subdirectories

**Review by jlesquembre — APPROVED** [review]

**Pull request merged**